sgxs-loaders/src/enclaveapi/mod.rs

```diff
@@ -0,0 +1,268 @@
+            &Init(Ret(ErrorCode::InvalidCpusvn)) |
+            &Init(Ret(ErrorCode::InvalidAttribute)) | // InvalidEinitAttribute according to PR, but does not exist.
+            &Init(Ret(ErrorCode::InvalidMeasurement)) => true,
+            _ => false,
+        }
+    }
+}
+
+impl EnclaveLoad for WinInnerLibrary {
+    type Error = Error;
+
+    fn new(
+        device: Arc<WinInnerLibrary>,
+        ecreate: MeasECreate,
+        attributes: Attributes,
+        miscselect: Miscselect,
+    ) -> Result<Mapping<Self>, Self::Error> {
+        let secs = Secs {
+            size: ecreate.size,
+            ssaframesize: ecreate.ssaframesize,
+            miscselect,
+            attributes,
+            ..Default::default()
+        };
+
+        let curhandle: HANDLE = unsafe {GetCurrentProcess()};
+        let base = unsafe {
+            CreateEnclave(
+                curhandle,
+                ptr::null_mut(),
+                ecreate.size as _,
+                0,
+                ENCLAVE_TYPE_SGX,
+                &secs as *const _ as *const c_void,
+                mem::size_of::<Secs>() as _,
+                ptr::null_mut()
+            )
+        };
+
+        if base.is_null() {
+            Err(Error::Create(IoError::last_os_error()))
+        } else {
+            Ok(Mapping {
+                device,
+                tcss: vec![],
+                base: base as _,
+                size: ecreate.size,
+            })
+        }
+    }
+    fn add(
+        mapping: &mut Mapping<Self>,
+        page: (MeasEAdd, PageChunks, [u8; 4096]),
+    ) -> Result<(), Self::Error> {
+
+        let (eadd, chunks, data) = page;
+
+        if eadd
+            .secinfo
+            .flags
+            .intersects(SecinfoFlags::PENDING | SecinfoFlags::MODIFIED | SecinfoFlags::PR)
+        {
+            return Err(Error::Add(ErrorKind::InvalidInput.into()));
+        }
+        let mut flags : winapi::shared::minwindef::DWORD = 0;
+        match (eadd.secinfo.flags & (SecinfoFlags::R | SecinfoFlags::W | SecinfoFlags::X)).bits() as u8 {
+            0b000 => {},
+            0b001 => flags = PAGE_READONLY,
+            0b010 => {return Err(Error::Add(ErrorKind::InvalidInput.into()))},
+            0b011 => flags = PAGE_READWRITE,
+            0b100 => flags = PAGE_EXECUTE,
+            0b101 => flags = PAGE_EXECUTE_READ,
+            0b110 => {return Err(Error::Add(ErrorKind::InvalidInput.into()))},
+            0b111 => flags = PAGE_EXECUTE_READWRITE,
+            0b1000...255 => unreachable!(),
+        }
+
+        match PageType::from_repr(eadd.secinfo.flags.page_type()) {
+            Some(PageType::Reg) => {}
+            Some(PageType::Tcs) => {
+                if eadd.secinfo.flags.contains(SecinfoFlags::R | SecinfoFlags::W | SecinfoFlags::X) {
+                    return Err(Error::Add(ErrorKind::InvalidInput.into()))
+                }
+                // NOTE: For some reason the windows API needs the Read flag set but then removes it
```

Comment:
I don't remember the exact details, but I believe it's like this, is that right? You *must* call `LoadEnclaveData` with `PAGE_ENCLAVE_THREAD_CONTROL|PAGE_READWRITE` but Windows will call EADD with only `SecinfoFlags::R`? If that is the case, the error checking here is not sufficient.

Reply:
I think you are right about the flag requirements for calling LoadEnclaveData. 
However, I am unsure of the following.
1) What are the flags that the windows loader should get for a `PageType::Tcs`?
what does the linux loader get? 
2) Where are these flags set? From what I see, they are defined in the `.sig` file. Where is the .sig file generated?
3) If the same `.sig` file should work on both linux and windows, then shouldn't the windows loader accept the same flags as the linux loader?

Reply:
When you say `.sig`, do you mean `.sgxs`?

Yes Windows should be accepting whatever flag we want, but it doesn't, and there's nothing we can do about it. So our only recourse is throwing an error if we know it won't work...

Reply:
Ok.
 I will change the error checking to check if `eadd.secinfo.flags` contains both the R & W flags, and if it doesn't then throw an error. 

However (I guess) this would mean that we would have to build sgxs files differently for windows.

Reply:
> However (I guess) this would mean that we would have to build sgxs files differently for windows.

Why? We already know it works right now.

Reply:
The way it was working was that if the PageType is Tcs , we were forcing the flags to be the following. 
```
line number 127: flags = PAGE_ENCLAVE_THREAD_CONTROL|PAGE_READWRITE;
```

In the sgxs file which was built on linux, the value of `eadd.secinfo.flags` for a `PageType::Tcs` was `PT_B0` when i tried to find it's value while loading it on windows.

Reply:
Yes, that's exactly the behavior I described above?